coremltools/converters/mil/mil/passes/defs/optimize_normalization.py

```diff
@@ -823,6 +830,163 @@ def _try_match_and_transform_pattern_4(self, reduce_op: Operation, block: Block)
+        if len(list(root_var.child_ops)) < 2:
+            return False
+        # Do not enforce that the only child ops are reduce_mean and sub as in other
+        # patterns. There are models where the root op is used after the layer norm.
+
+        # check 1st reduce_mean op
+        if not self._check_reduce_op(reduce_op):
+            return False
+        ops_to_remove.append(reduce_op)
+
+        # check 1st sub op
+        if not self._check_child_op_types(reduce_op, ["sub"], check_order=False):
+            return False
+        child_ops_reduce_mean = list(reduce_op.outputs[0].child_ops)
+        sub_op1 = child_ops_reduce_mean[0]
+        if not (sub_op1.x == root_var and sub_op1.y == reduce_op.outputs[0]):
+            return False
+        ops_to_remove.append(sub_op1)
+
+        # check mul op (equivalent to a square op)
+        square_op = self._try_get_child_op_type(sub_op1, "mul")
+        if square_op is None:
+            return False
+        if square_op.x != square_op.y:
+            return False
+        ops_to_remove.append(square_op)
+
+        # check second reduce mean
+        reduce_op2 = self._try_get_child_op_type(square_op, "reduce_mean")
+        if not self._check_reduce_op(reduce_op2):
+            return False
+        ops_to_remove.append(reduce_op2)
+
+        # check add op (with epsilon)
+        add_op1 = self._try_get_child_op_type(reduce_op2, "add")
+        if add_op1 is None:
+            return False
+        epsilon_var = add_op1.y if add_op1.x == reduce_op2.outputs[0] else add_op1.x
+        if epsilon_var.val is None or len(epsilon_var.val.shape) != 0:
+            return False  # must be scalar
+        ops_to_remove.append(add_op1)
+
+        # check rsqrt
+        rsqrt_op = self._try_get_child_op_type(add_op1, "rsqrt")
+        if rsqrt_op is None:
+            return False
+        ops_to_remove.append(rsqrt_op)
+
+        ## without elementwise affine
+        mul_op = self._try_get_child_op_type(rsqrt_op, "mul")
+        if mul_op is None:
+            return False
+        if mul_op.y != sub_op1.outputs[0] and mul_op.x != sub_op1.outputs[0]:
+            return False
+        ops_to_remove.append(mul_op)
+
+        # Default values if no gamma or beta ops.
+        end_op = mul_op
+        gamma_var = None
+        beta_var = None
+
+        add_beta_op = self._try_get_child_op_type(mul_op, "add")
+        mul_gamma_op = self._try_get_child_op_type(add_beta_op, "mul")
+
+        if add_beta_op is not None and mul_gamma_op is not None:
+            beta_var = add_beta_op.y if add_beta_op.x == mul_op.outputs[0] else add_beta_op.x
+
+            gamma_var = mul_gamma_op.y if mul_gamma_op.x == add_beta_op.outputs[0] else mul_gamma_op.x
+            gamma_var = mb.const(
+                val=np.squeeze(gamma_var.val),
+                name="_fuse_layernorm_gamma",
+            )
+
+            # Scale beta by gamma. Note: this un-scaling introduces a little error.
```

Comment:
a little error you mean numerical precision loss?

Reply:
Yes I think there's some precision lost due to the floating point math. This isn't my strong suit, so could be wrong.

```python
>>> import torch
# distilbert_{weight,bias} as they are from huggingface
>>> distilbert_weight = torch.randn(4000)
>>> distilbert_bias = torch.randn(4000)
# ane_ln_bias as adjusted here: https://github.com/apple/ml-ane-transformers/blob/da64000fa56cc85b0859bc17cb16a3d753b8304a/ane_transformers/huggingface/distilbert.py#L31-L33
>>> ane_ln_bias = distilbert_bias / distilbert_weight
# coremltools_ln_bias as adjusted in this PR
>>> coremltools_ln_bias = ane_ln_bias * distilbert_weight
>>> torch.equal(coremltools_ln_bias, distilbert_bias)
False
>>> torch.allclose(coremltools_ln_bias, distilbert_bias)
True
```